ext/mixed/js/object-property-accessor.js

```diff
@@ -0,0 +1,244 @@
+ */
+
+/**
+ * Class used to get and set generic properties of an object by using path strings.
+ */
+class ObjectPropertyAccessor {
+    constructor(target, setter=null) {
+        this._target = target;
+        this._setter = (typeof setter === 'function' ? setter : null);
+    }
+
+    getProperty(pathArray, pathLength) {
+        let target = this._target;
+        const ii = typeof pathLength === 'number' ? Math.min(pathArray.length, pathLength) : pathArray.length;
+        for (let i = 0; i < ii; ++i) {
+            const key = pathArray[i];
+            if (!ObjectPropertyAccessor.hasProperty(target, key)) {
+                throw new Error(`Invalid path: ${ObjectPropertyAccessor.getPathString(pathArray.slice(0, i + 1))}`);
+            }
+            target = target[key];
+        }
+        return target;
+    }
+
+    setProperty(pathArray, value) {
+        if (pathArray.length === 0) {
+            throw new Error('Invalid path');
+        }
+
+        const target = this.getProperty(pathArray, pathArray.length - 1);
+        const key = pathArray[pathArray.length - 1];
+        if (!ObjectPropertyAccessor.isValidPropertyType(target, key)) {
+            throw new Error(`Invalid path: ${ObjectPropertyAccessor.getPathString(pathArray)}`);
+        }
+
+        if (this._setter !== null) {
+            this._setter(target, key, value, pathArray);
+        } else {
+            target[key] = value;
+        }
+    }
+
+    static getPathString(pathArray) {
+        const regexShort = /^[a-zA-Z_][a-zA-Z0-9_]*$/;
+        let pathString = '';
+        let first = true;
+        for (let part of pathArray) {
+            switch (typeof part) {
+                case 'number':
+                    if (Math.floor(part) !== part || part < 0) {
+                        throw new Error('Invalid index');
+                    }
+                    part = `[${part}]`;
+                    break;
+                case 'string':
+                    if (!regexShort.test(part)) {
+                        const escapedPart = part.replace(/["\\]/g, '\\$&');
+                        part = `["${escapedPart}"]`;
+                    } else {
+                        if (!first) {
+                            part = `.${part}`;
+                        }
+                    }
+                    break;
+                default:
+                    throw new Error(`Invalid type: ${typeof part}`);
+            }
+            pathString += part;
+            first = false;
+        }
+        return pathString;
+    }
+
+    static getPathArray(pathString) {
+        const pathArray = [];
+        let state = 0;
+        let quote = 0;
+        let value = '';
+        let escaped = false;
+        for (const c of pathString) {
+            const v = c.codePointAt(0);
+            switch (state) {
+                case 0: // Empty
+                case 1: // Expecting identifier start
```

Comment:
Do you think the states could be class variables instead? It would require putting them at bottom, but variable names are easier to read than plain integers especially when the state is changed from different places.

Reply:
So for class variables, the recommendation is to use strings instead of "enum" constants, as per https://github.com/FooSoft/yomichan/pull/157#discussion_r257526239. However, I chose to not do that in this case because there would be added overhead for string comparisons in the `switch`, which would occur for every character of the string.

IMO since the state integers are completely self-contained within the function, and the cases are commented, this should be sufficiently explanatory and maximally efficient. I could also comment the places where `state` is assigned with the same comment if you think that would be helpful.

Reply:
Ok, after reading the earlier comments I understand the reason why class enums don't work so well in JS. Another option worth considering would be creating int-backed local constants for the states. Also, did you test how much it would affect performance to use strings for states?

Reply:
> I could also comment the places where `state` is assigned with the same comment if you think that would be helpful.

I think that could be too difficult to maintain because the code would work just fine with incomplete or incorrect comments.

Reply:
> Another option worth considering would be creating int-backed local constants for the states. Also, did you test how much it would affect performance to use strings for states?

So I have just tested both of these using https://jsbench.me/ on Firefox, Chrome, and Edge, and these are the results:
* On Chrome, using direct integers, strings, or integer constants are all equally fast. v8 may have a good optimization for this which is causing the comparable speeds.
* On Firefox, direct integers are the fastest, strings are about 8% slower, and integer constants are about 4% slower.
* On Edge, direct integers are the fastest, strings are maybe 10% slower, and integer constants are roughly the same as direct integers if the constants are declared outside the function body.
* Chrome speed > Firefox speed > Edge speed.

So based on this, I could probably just switch to using strings. I was expecting the speed loss of string comparisons to be more significant.

Reply:
Thanks a lot for testing! Did you test using this function or just comparisons? I got similar results testing a for loop with simple `===` checks, but realized that browsers might optimize those differently.

Reply:
https://gist.github.com/toasted-nutbread/366a15c1b4e93ddf656f44e184d439d2

Switch-case can be optimized better than `if (...) else if (...) else if (...) ...` when comparing the same var, and switch-case uses strict equality.

Reply:
Update to strings in 4b699a6b46869d5766f331d863aea38374ece50a